### ISSUE TITLE
Optimize AOpCodeTable.GetInst

### ISSUE THREAD

**Issue by gdkchan** [Ryujinx/Ryujinx#282]
Currently, it does a linear search and in the worst case, it will need to compare the opcode value with the value on all the (currently) 837 entries on the opcode table, which is done using a linked list. This number will keep increasing as new instructions are added and it is only going to get slower, so ideally we should find a more efficient way to find instructions on the table from the opcode value.

**Comment by strubelz** [comment]
I have an idea about some kind of search tree structure, I'll check it out tommorow. And I think it should be faster to read from from an ArrayList or Array.

**Comment by strubelz** [comment]
Hmm, I can't think of any solution using a hash table. The entire point of a hash is that is changes completly when one bit of the source is changed. For me it seems pretty hard to do some kind of pattern matching with that, but you guys know more than me.

**Comment by emmauss** [comment]
well, i thought that with some kind of sorted list or dictionary, we could do a binary search or a hash lookup(dictionary uses those, i think). though this is feasible if the expected value of an instruction does not change, and set is not called after the table has been populated.

**Comment by mailwl** [comment]
just an idea - in most cases, only 22bits of opcode is mean for differ operation, so, it's possibly use array of 2^22 elements (32mb?) to find needed opcode in one step, something like
`instr[opcode & 0x003FFFFF]`

**Comment by MerryMage** [comment]
I agree with mailwl, though personally I think only 8 bits are necessary (256 entries for ~800 instruction types), and preferably the most-significant bits.

**Comment by mailwl** [comment]
@MerryMage made good solution, i think thre is no need further optimizations (at least, before the actuil time used for fetching instructions in whole jit compiling)
